packages/backend-common/src/service/ServiceBuilderImpl.ts

```diff
@@ -0,0 +1,117 @@
+/*
+ * Copyright 2020 Spotify AB
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import compression from 'compression';
+import cors from 'cors';
+import express, { Router } from 'express';
+import helmet from 'helmet';
+import { Server } from 'http';
+import { Logger } from 'winston';
+import { getRootLogger } from '../logging';
+import {
+  errorHandler,
+  notFoundHandler,
+  requestLoggingHandler,
+} from '../middleware';
+import { ServiceBuilder } from './types';
+
+const DEFAULT_PORT = 7000;
+
+export class ServiceBuilderImpl implements ServiceBuilder {
+  private port: number | undefined;
+  private logger: Logger | undefined;
+  private corsOptions: cors.CorsOptions | undefined;
+  private routers: [string, Router][];
+
+  constructor() {
+    this.routers = [];
```

Comment:
I'd rather go for a `public app:Express;` and here create an empty one and on `addRouter` just do `this.app.use`. Also, maybe add `addMiddleware`. The reason behind is `start` being too smart now, ideally it should be just this.app.listen(). This maybe also involves moving all the predefined middlewares in the constructor

Reply:
I preferred to do the work in `build` i guess because of java habits - i like the builder to be extremely dumb and being just an immutable data class. That way, you can construct the entire builder using something like AutoMatter essentially and then you can `build` over and over again without problem, and even place the `build` logic externally if you prefer. See what I mean? The builder is almost more an immutable `FooConfig` that happens to be buildable, than a `FooBuilder` that is more lax and mutable and has to be thrown away (i.e. you have to use it the right way, because if you don't you cause breakage).